File: playService/assets/src/game/game-ui.js
const _ = require('lodash')
const h = require('snabbdom/h').default
const constants = require('../common/constants')
const button = require('../common/button')
const modal = require('../common/modal')
const gameBoard = require('./game-board')
const images = require('../common/images')


function render(state, actionHandler) {
  let m = null
  const closeAction = [actionHandler, {closeModal: true}]
  if (state.modal === 'log')
    m = modal(renderLog(state.game.events), closeAction)
  else if (state.modal === 'playerList')
    m = modal(renderPlayerList(state), closeAction)
  else if (state.modal && state.modal.type === 'previewScenario')
    m = modal(h('div.modal-maximized',
        gameBoard.renderCanvas(state.modal.scenario, state.modal.scenario.initialRobots.map(gameBoard.robotFromInitial), {}),
        ),
        closeAction)

  const game = state.game
  const playerIndex = _.get(game.players.find(p => p.name === state.player), "index")
  return h('div.game', [
    fab('.fab-right-1', images.iconClose, [actionHandler, {leaveGame: true}]),
    fab('.fab-left-1', images.iconReplayAnimation, [actionHandler, {replayAnimations: state.animations}]),
    fab('.fab-left-2', playerIndex !== undefined ? images.player(playerIndex) : images.iconGamerlist, [actionHandler, {setModal: 'playerList'}]),
    h('div.game-board', gameBoard.renderCanvas(game.scenario, game.players.map(gameBoard.robotFromPlayer), {
      animationStart: state.animationStart,
      frames: state.animations
    })),
    renderActionButtons(state, game, actionHandler),
    m])
}

function fab(classes, image, onclick){
  return h('div.fab'+classes, {on: {click: onclick}},
      h('img', {props: {src: image.src}}))
}

function renderPlayerList(state) {
  let rows = state.game.players.map(function (player) {
    return h('tr', [
      h('td', h('img', {
        props: {src: images.player(player.index).src},
        style: {'max-width': '20px', 'max-height': '20px'}
      })),
      h('td', player.name),
      h('td', player.finished ? 'finished as ' + player.finished.rank : (player.ready || _.get(player.instructions, 'length', 0) ? 'ready' : '')),
    ])
  })

  const header = h('tr', [h('th', ''), h('th', 'name'), h('th', 'state')])

  return h('div', [
    h('h4', 'Players: '),
    h('table', [header, ...rows])
  ])
}

function renderActionButtons(state, game, actionHandler) {
  const focusedSlot = state.focusedSlot || 0
  const player = game.players.find((player) => player.name === state.player)

  function instructionCard(type) {
    function unusedAndThisType(opt, index) {
      return opt[type] && !_.some(player.instructionSlots, slot => slot === index)
    }

    const image = images.action(type)
    const count = player.instructionOptions.filter(unusedAndThisType).length
    const unusedIndex = _.findIndex(player.instructionOptions, unusedAndThisType)
    return h('div.action',
        {on: {click: count !== 0 ? () => actionHandler({setInstruction: true, slot: focusedSlot, instruction: unusedIndex}) : undefined}},
        [h('img', {props: {src: image.src}}), h('div.badge', count)])
  }

  function instructionSlot(index) {
    const instruction = player.instructionSlots[index]
    const focused = focusedSlot === index
    const props = {
      class: {"slot-focused": focused},
      on: {
        dblclick: () => actionHandler({resetSlot: true, slot: index}),
        click: () => actionHandler({focusSlot: index})
      }
    }
    if (instruction !== undefined && instruction !== null) {
      const image = images.action(Object.keys(player.instructionOptions[instruction])[0])
      return h('div.slot.slot-filled', props,
          h('img', {props: {src: image.src}}))
    } else {
      return h('div.slot', props, index+1)
    }
  }

  if (!player && game.cycle === 0) {
    return h('div.footer-group', [
      h('div.slots-panel', 'observer mode'),
      h('div.cards-panel', button.builder.primary()(actionHandler, {joinGame: state.gameId}, 'Join Game'))
    ])
  } else if (!player) {
    return h('div.status-panel', h('div.text', 'observer mode'))
  } else if (player.finished) {
    return h('div.status-panel', h('div.text', 'target reached'))
  } else {
    let instructionTypes = []
    let instr = _.clone(player.instructionOptions)
    while(instr.length !== 0) {
      let head = instr[0]
      let type = Object.keys(head)[0]
      instructionTypes.push(type)
      instr = _.dropWhile(instr, i => i[type])
    }

    return h('div.footer-group', [
      h('div.slots-panel', _.range(constants.numberOfInstructionsPerCycle).map(instructionSlot)),
      h('div.cards-panel', instructionTypes.map(instructionCard))
    ])
  }
}

function renderLog(events) {
  return h('div', [
    h('h4', 'Log: '),
    h('div', events.map(log => h('div', log)))
  ])
}

module.exports = render

File: frontend/src/main/scala/roborace/frontend/pages/multiplayer/game/GameUi.scala
package roborace.frontend.pages
package multiplayer.game

import scala.concurrent.ExecutionContext.Implicits.global

import entities._
import logic.command.{DeregisterForGame, RegisterForGame}
import org.scalajs.dom
import roborace.frontend.pages.components._
import roborace.frontend.pages.components.gameBoard.{Animation, RenderGame}
import roborace.frontend.pages.multiplayer.game.GamePage.Context
import roborace.frontend.pages.multiplayer.lobby.LobbyPage
import roborace.frontend.service.{Actions, Service, ServiceTrait}
import roborace.frontend.util.{SnabbdomEventListener, Untyped}
import snabbdom.{Event, Node, Snabbdom}
object GameUi extends snabbdom.Syntax {
  def apply(implicit context: Context): Node = {
    context.local.game.entity.players.find(_.id == context.global.sessionId) match {
      case None if context.local.game.entity.cycle == 0 =>
        Body
          .game()
          .child(returnToLobbyFab())
          .child(replayFab())
          .child(openPlayerListFab())
          .child(syncFab())
          .child(
            RenderGame(
              context.local.game,
              Some { (p, _) =>
                context.local.game.entity.scenario.initialRobots
                  .find(r => r.position == p && !context.local.game.entity.players.contains(r))
                  .foreach(r => Actions.sendCommand(RegisterForGame(r.index)))
              }
            )
          )
          .child(
            Node("div").classes("text-panel").text("tap a start position to join the game")
          )
          .childOptional(playerListModal())

      case None =>
        Body
          .game()
          .child(returnToLobbyFab())
          .child(replayFab())
          .child(openPlayerListFab())
          .child(syncFab())
          .child(RenderGame(context.local.game, None))
          .child(Node("div.text-panel").text("observer mode"))
          .childOptional(playerListModal())

      case Some(you: QuitedPlayer) =>
        Body
          .game()
          .children(
            returnToLobbyFab(),
            replayFab(),
            openPlayerListFab(),
            syncFab(),
            RenderGame(context.local.game, None),
            Node("div.text-panel").text("game quitted")
          )
          .childOptional(playerListModal())

      case Some(you: FinishedPlayer) =>
        Body
          .game()
          .children(
            returnToLobbyFab(),
            replayFab(),
            openPlayerListFab(),
            syncFab(),
            RenderGame(context.local.game, None),
            Node("div.text-panel").text("target reached as " + you.rank)
          )
          .childOptional(playerListModal())

      case Some(you: RunningPlayer) =>
        val color = RobotColor.dark(you.index)
        Body
          .game()
          .style("--highlight-color", color)
          .children(
            Fab(Icons.close)
              .classes("fab-right-1")
              .event("click", SnabbdomEventListener.sideeffect(() => Actions.sendCommand(DeregisterForGame))),
            openPlayerListFab(),
            syncFab(),
            replayFab(),
            RenderGame(context.local.game, None),
            instructionSlots(you),
            cardsBar(you)
          )
          .childOptional(playerListModal())
    }
  }

  private def replayFab()(implicit context: Context) =
    Fab(Icons.replay)
      .classes("fab-left-1")
      .event(
        "click",
        (_: Event) => {
          Untyped(dom.document.querySelector(".game-board svg")).setCurrentTime(
            if (context.local.game.entity.cycle == 0)
              0
            else
              Animation.eventSequenceDuration(context.local.game.entity.events.takeWhile {
                case s: StartCycleEvaluation => s.cycle != context.local.game.entity.cycle - 1
                case _                       => true
              })
          )
        }
      )
      .event("dblclick", SnabbdomEventListener.sideeffect(() => Untyped(dom.document.querySelector(".game-board svg")).setCurrentTime(0)))

  private def syncFab()(implicit context: Context): Node =
    Fab(Icons.sync)
      .classes("fab-right-2")
      .event[Event](
        "click",
        _ =>
          Service
            .getGame(context.local.game.id)
            .map(GameState.game.set)
            .foreach(f => context.update(f(context.local)))
      )

  private def openPlayerListFab()(implicit context: Context) =
    Fab(Icons.list)
      .classes("fab-left-2")
      .event("click", SnabbdomEventListener.modify(GameState.playerModalOpened.set(true)))

  private def returnToLobbyFab()(implicit context: Context) =
    Fab(Icons.close).classes("fab-right-1").event("click", SnabbdomEventListener.set(LobbyPage.load()))

  def playerListModal()(implicit context: Context): Option[Node] =
    if (context.local.playerModalOpened) {
      Some(
        Modal(closeAction = SnabbdomEventListener.modify(GameState.playerModalOpened.set(false)))(
          Node("h1.title").text("Player List"),
          Node("table.table.is-fullwidth")
            .child(
              "tr"
                .child("th".text("Player"))
                .child("th".text("State"))
            )
            .child(
              context.local.game.entity.players.map(
                player =>
                  "tr"
                    .child("td".child("img.image.is-64x64".attr("src", Images.player(player.index))))
                    .child(
                      "td".text(
                        player match {
                          case player: RunningPlayer if player.instructionSlots.isEmpty => "waiting for"
                          case _: RunningPlayer                                         => "ready"
                          case _: QuitedPlayer                                          => "left the game"
                          case player: FinishedPlayer                                   => s"finished as ${player.rank}"
                        }
                      )
                    )
              )
            )
        )
      )
    } else {
      None
    }

  def instructionSlots(player: RunningPlayer)(implicit context: Context) = {
    def instructionSlot(index: Int): Node = {
      val instruction = context.local.slots.get(index)
      val focused     = context.local.focusedSlot == index

      val setFocus  = SnabbdomEventListener.modify(GameState.focusedSlot.set(index))
      val resetSlot = SnabbdomEventListener.sideeffect(() => Actions.unsetInstruction(index))

      instruction match {
        case Some(instr) =>
          Node("span.slot.filled")
            .`class`("focused", focused)
            .event("click", setFocus)
            .event("dblclick", resetSlot)
            .child(Node("img").attr("src", Images.instructionIcon(instr)))
        case None =>
          Node("span.slot").`class`("focused", focused).event("click", setFocus).text((index + 1).toString)
      }
    }

    Node("div.nowrap-panel").child((0 until Constants.instructionsPerCycle).map(instructionSlot))
  }

  private def cardsBar(player: RunningPlayer)(implicit context: Context): Node = {
    def instructionCard(instruction: Instruction): Option[Node] = {
      val allowed = player.instructionOptions.find(_.instruction == instruction).fold(0)(_.count)
      val used    = context.local.slots.values.count(_ == instruction)
      val free    = allowed - used

      if (free > 0) {
        Some(
          Node("div.instruction-card")
            .classes(s"stacked-instruction-card-${free.min(5)}")
            .child(
              Node("img")
                .attr("src", Images.instructionIcon(instruction))
            )
            .childOptional(
              if (free != 1) Some(Node("div.badge").text(free.toString))
              else None
            )
            .event("click", SnabbdomEventListener.sideeffect(() => Actions.placeInstruction(instruction, context.local.focusedSlot)))
        )
      } else {
        None
      }
    }

    Node("div.nowrap-panel").child(Instruction.instructions.flatMap(instructionCard))
  }

}
